src/librustc/traits/error_reporting.rs

```diff
@@ -103,16 +105,191 @@ pub fn report_projection_error<'a, 'tcx>(infcx: &InferCtxt<'a, 'tcx>,
+}
+
+trait AssociatedWeight {
+    fn get_weight(&self) -> usize;
+}
+
+impl<'a> AssociatedWeight for TypeVariants<'a> {
+    // First number is for "global" weight and second number is for bigger precision
+    fn get_weight(&self) -> usize {
+        match *self {
+            TypeVariants::TyBool => 11,
+            TypeVariants::TyChar => 12,
+            TypeVariants::TyStr => 13,
+            TypeVariants::TyInt(_) => 21,
+            TypeVariants::TyUint(_) => 22,
+            TypeVariants::TyFloat(_) => 23,
+            TypeVariants::TyRawPtr(_) => 24,
+            TypeVariants::TyEnum(_, _) => 31,
+            TypeVariants::TyStruct(_, _) => 32,
+            TypeVariants::TyBox(_) => 33,
+            TypeVariants::TyTuple(_) => 34,
+            TypeVariants::TyArray(_, _) => 41,
+            TypeVariants::TySlice(_) => 42,
+            TypeVariants::TyRef(_, _) => 51,
+            TypeVariants::TyFnDef(_, _, _) => 52,
+            TypeVariants::TyFnPtr(_) => 53,
+            TypeVariants::TyTrait(_) => 61,
+            TypeVariants::TyClosure(_, _) => 71,
+            TypeVariants::TyProjection(_) => 81,
+            TypeVariants::TyParam(_) => 82,
+            TypeVariants::TyInfer(_) => 83,
+            TypeVariants::TyError => 91,
+        }
+    }
+}
+
+// The "closer" the types are, the lesser the weight
+fn get_weight_diff(a: &ty::TypeVariants, b: &TypeVariants, big_weight: bool) -> usize {
+    let w1 = if big_weight { a.get_weight() / 10 } else { a.get_weight() % 10 };
+    let w2 = if big_weight { b.get_weight() / 10 } else { b.get_weight() % 10 };
+
+    if w1 < w2 {
+        w2 - w1
+    } else {
+        w1 - w2
+    }
+}
+
+// Once we have "globally matching" types, we need to run another filter on them
+fn filter_matching_types<'tcx>(weights: &[(usize, usize)],
+                               imps: &[(DefId, subst::Substs<'tcx>)],
+                               trait_types: &[ty::Ty<'tcx>])
+                               -> usize {
+    let matching_weight = weights[0].1;
+    let iter = weights.iter().filter(|&&(_, weight)| weight == matching_weight);
+    let mut filtered_weights = vec!();
+
+    for &(pos, _) in iter {
+        let mut weight = 0;
+        for (type_to_compare, original_type) in imps[pos].1
+                                                         .types
+                                                         .get_slice(ParamSpace::TypeSpace)
+                                                         .iter()
+                                                         .zip(trait_types.iter()) {
+            weight += get_weight_diff(&type_to_compare.sty, &original_type.sty, false);
+        }
+        filtered_weights.push((pos, weight));
+    }
+    filtered_weights.sort_by(|a, b| a.1.cmp(&b.1));
+    filtered_weights[0].0
+}
+
+fn get_best_matching_type<'tcx>(imps: &[(DefId, subst::Substs<'tcx>)],
+                                trait_types: &[ty::Ty<'tcx>]) -> usize {
+    let mut weights = vec!();
```

Comment:
Why not use `u32` for the weight instead of `usize` ?

As I understand it, here `weights` has type `Vec<(usize, usize)>`, where the `.0` is the index of some impl within `imps` (lets call it *I*), and the `.1` is the sum of the differences (in weight) of the types found via point-wise comparison between the types of *I* and the types of `original_type`.

However, *I* personally would be worried about accidentally mixing up the `.0` and the `.1` in the tuple -- they both have type `usize`, so we're missing out on some potential checking that we could get effectively for free by using (IIUC) a `u32` for the weight.